yt_dlp/extractor/rambler.py

```diff
@@ -0,0 +1,123 @@
+import json
+
+from .common import InfoExtractor
+from ..utils import (
+    ExtractorError,
+    base_url,
+    clean_html,
+    float_or_none,
+    join_nonempty,
+    js_to_json,
+    parse_iso8601,
+    parse_qs,
+    url_or_none,
+)
+from ..utils.traversal import require, traverse_obj
+
+
+class RamblerBaseIE(InfoExtractor):
+    def _extract_from_rambler_api(self, rambler_id, referrer):
+        key = 'uuid' if rambler_id.startswith('record::') else 'id'
+
+        player_data = self._download_json(
+            'https://api.vp.rambler.ru/api/v3/records/getPlayerData',
+            rambler_id, query={
+                'params': json.dumps({
+                    'checkReferrerCount': True,
+                    'referrer': referrer,
+                    key: rambler_id,
+                }).encode(),
+            })
+        if not player_data.get('success'):
+            error_type = traverse_obj(player_data, ('error', 'type', {str}, filter))
+            error_subtype = traverse_obj(player_data, ('error', 'subtype', {str}, filter))
+            raise ExtractorError(
+                join_nonempty(error_type, error_subtype, delim=': '), expected=True)
+        playlist = player_data['result']['playList']
+
+        formats = []
+        for m3u8_url in traverse_obj(playlist, (
+            ('directSource', 'source'), {url_or_none}, filter,
+        )):
+            formats.extend(self._extract_m3u8_formats(
+                m3u8_url, rambler_id, 'mp4', fatal=False))
+        self._remove_duplicate_formats(formats)
+
+        return {
+            'channel': traverse_obj(player_data, ('channel', {str})),
+            'formats': formats,
+            **traverse_obj(playlist, {
+                'id': ('uuid', {str}),
+                'title': ('title', {clean_html}),
+                'duration': ('duration', {float_or_none(scale=1000)}),
+                'release_timestamp': ('startsAt', {parse_iso8601}),
+                'thumbnail': (('customScreenshotOrig', 'snapshot'), {url_or_none}, any),
+                'webpage_url': ('defaultShareUrl', {url_or_none}),
+            }),
+        }
+
+
+class RamblerIE(RamblerBaseIE):
+    IE_NAME = 'rambler'
+    IE_DESC = 'Рамблер'
+
+    _VALID_URL = [
+        r'https?://(?:api\.)?vp\.rambler\.ru/(?:api/(?:other|v3)/)?(?:player/(?:embed|export)\.html|records/getPlayerData)',
```

Comment:
is there a reason for supporting API URLs? would a user typically have one of these URLs to pass without using dev tools?